File: YaAlgoTrainings/Training5.0/LinSearch/J.cpp
#include <iostream>
#include <vector>
#include <string>
#include <tuple>


using namespace std;

struct Rect {
    int frow;
    int fcol;
    int lrow;
    int lcol;

    bool operator!=(const Rect& r2) {
        return tie(frow, fcol, lrow, lcol) != tie(r2.frow, r2.fcol, r2.lrow, r2.lcol);
    }
};
const Rect NONE = {-1, -1, -1, -1};
using Sector = Rect;

bool isInside(int row, int col, Rect r) {
    return r.frow <= row && row <= r.lrow &&
        r.fcol <= col && col <= r.lcol;
}

Rect rectInSector(const vector<string>& picture, const Sector& s) {
    Rect r = NONE;
    for (int row = s.frow; row <= s.lrow; ++row) {
        for (int col = s.fcol; col <= s.lcol; ++col) {
            if (picture[row][col] == '#' && !isInside(row, col, r)) {
                if (r != NONE) {
                    return NONE;
                }
                r.frow = r.lrow = row; r.fcol = r.lcol = col;
                
                while (r.lrow + 1 <= s.lrow && picture[r.lrow + 1][col] == '#') {
                    r.lrow++;
                }

                while (r.lcol + 1 <= s.lcol && picture[row][r.lcol + 1] == '#') {
                    r.lcol++;
                }
            }
        }
    }

    for (int row = r.frow + 1; row <= r.lrow; ++row) {
        for (int col = r.fcol + 1; col <= r.lcol; ++col) {
            if (picture[row][col] != '#') {
                return NONE;
            }
        }
    }

    return r;
}

void fillRectangle(vector<string>& picture, const Rect& r, char c) {
    for (int row = r.frow; row <= r.lrow; ++row) {
        for (int col = r.fcol; col <= r.lcol; ++col) {
            picture[row][col] = c;
        }
    }
}

bool checkSectors(int m, int n, vector<string>& picture, const Sector& s1, const Sector& s2) {
    Rect r1 = rectInSector(picture, s1);
    Rect r2 = rectInSector(picture, s2);

    if (r1 != NONE && r2 != NONE) {
        fillRectangle(picture, r1, 'a');
        fillRectangle(picture, r2, 'b');
        return true;
    } else {
        return false;
    }
}

bool isTwoRect(int m, int n, vector<string>& picture) {
    for (int row = 0; row + 1 < m; ++row) {
        Sector s1 = {0, 0, row, n - 1};
        Sector s2 = {row + 1, 0, m - 1, n - 1};
        if (checkSectors(m, n, picture, s1, s2)) {
            return true;
        }
    }

    for (int col = 0; col + 1 < n; ++col) {
        Sector s1 = {0, 0, m - 1, col};
        Sector s2 = {0, col + 1, m - 1, n - 1};
        if (checkSectors(m, n, picture, s1, s2)) {
            return true;
        }
    }

    return false;
}

int main() {
    int m, n;
    cin >> m >> n;
    vector<string> picture(m);
    for (string& line : picture) {
        cin >> line;
    }

    if (isTwoRect(m, n, picture)) {
        cout << "YES\n";
        for (const string& line : picture) {
            cout << line << "\n";
        }
    } else {
        cout << "NO\n";
    }

    return 0;
}

File: YaAlgoTrainings/Training5.0/LinSearch/J.py
"""
Description:
mxn rectangle field
there is a picture on the field
check if the picture can be represented as 2 black rectangles without intersection
INPUT:
m n <= 200
description of field "." - blank, "#" - black cell
OUTPUT
YES
    field with a for first rectangle and b for second if it's possible
NO otherwise

Solution:
m n <= 200
algorithm 2 * 200^3
iterate through rows or columns as delimeters between two sectors
Identify if there is only one rectangle in each sector
"""
import unittest

def isRectangle(picture, rect):
    if rect is None:
        return False
    
    (frow, fcol), (lrow, lcol) = rect
    if (lrow - frow + 1) * (lcol - fcol + 1) <= 0:
        return False

    for row in range(frow, lrow + 1):
        for col in range(fcol, lcol + 1):
            if picture[row][col] != '#':
                return False
    
    return True

def fillRectangle(picture, rect, filler):
    (frow, fcol), (lrow, lcol) = rect

    for row in range(frow, lrow + 1):
        for col in range(fcol, lcol + 1):
            picture[row][col] = filler

def isInside(row, col, rectCand):
    if rectCand is None:
        return False
    
    (frow, fcol), (lrow, lcol) = rectCand
    return frow <= row <= lrow and fcol <= col <= lcol

def rectInSector(picture, sector):
    (frow, fcol), (lrow, lcol) = sector
    rectCand = None
    for row in range(frow, lrow + 1):
        for col in range(fcol, lcol + 1):
            if picture[row][col] == '#' and not isInside(row, col, rectCand):
                if rectCand is not None:
                    return None
                first = (row, col)
                lastRow = row
                lastCol = col
                while lastRow + 1 <= lrow and picture[lastRow + 1][col] == '#':
                    lastRow += 1
                
                while lastCol + 1 <= lcol and picture[row][lastCol + 1] == '#':
                    lastCol += 1
                last = (lastRow, lastCol)
                rectCand = (first, last)
    return rectCand if isRectangle(picture, rectCand) else None

def checkSectors(m, n, picture, sec1, sec2):
    rect1 = rectInSector(picture, sec1)
    rect2 = rectInSector(picture, sec2)
    if rect1 is not None and rect2 is not None:
        fillRectangle(picture, rect1, 'a'); fillRectangle(picture, rect2, 'b')
        return True
    else:
        return False

def isTwoRect(m, n, picture):
    for row in range(m - 1):
        sec1 = ((0, 0), (row, n - 1))
        sec2 = ((row + 1, 0), (m - 1, n - 1))
        if checkSectors(m, n, picture, sec1, sec2):
            return True

    for col in range(n - 1):
        sec1 = ((0, 0), (m - 1, col))
        sec2 = ((0, col + 1), (m - 1, n - 1))
        if checkSectors(m, n, picture, sec1, sec2):
            return True
    
    return False


def main():
    m, n = (int(x) for x in input().split())
    picture = [list(input()) for _ in range(m)]

    if isTwoRect(m, n, picture):
        print("YES")
        for rowItems in picture:
            print("".join(str(x) for x in rowItems))
    else:
        print("NO")


class J_Test(unittest.TestCase):
    pass


if __name__ == "__main__":
    main()